src/wazuh_modules/syscollector/syscollector_windows.c

```diff
@@ -24,49 +43,47 @@ hw_info *get_system_windows();
 /* From process ID get its name */
 
 char* get_process_name(DWORD pid){
-
-    char *string = NULL;
-    FILE *output;
-    char *command;
-    char *end;
     char read_buff[OS_MAXSTR];
-    int status;
-
-    memset(read_buff, 0, OS_MAXSTR);
-    os_calloc(COMMAND_LENGTH, sizeof(char), command);
-    snprintf(command, COMMAND_LENGTH - 1, "wmic process where processID=%lu get Name", pid);
-    output = popen(command, "r");
-    if (!output) {
-        mtwarn(WM_SYS_LOGTAG, "Unable to execute command '%s'.", command);
-    } else {
-        if (fgets(read_buff, OS_MAXSTR, output)) {
-            if (strncmp(read_buff,"Name", 4) == 0) {
-                if (!fgets(read_buff, OS_MAXSTR, output)){
-                    mtwarn(WM_SYS_LOGTAG, "Unable to get process name.");
-                    string = strdup("unknown");
-                }
-                else if (end = strpbrk(read_buff,"\r\n"), end) {
-                    *end = '\0';
-                    int i = strlen(read_buff) - 1;
-                    while(read_buff[i] == 32){
-                        read_buff[i] = '\0';
-                        i--;
-                    }
-                    string = strdup(read_buff);
-                }else
-                    string = strdup("unknown");
-            }
-        } else {
-            mtdebug1(WM_SYS_LOGTAG, "Unable to get process name (bad header).");
-            string = strdup("unknown");
-        }
-
-        if (status = pclose(output), status) {
-            mtwarn(WM_SYS_LOGTAG, "Command 'wmic' returned %d getting process ID.", status);
-        }
-    }
-
-    return string;
+	char *string = NULL, *ptr = NULL;
+	
+	/* Check if we are dealing with a system process */
+	if (pid == 0 || pid == 4)
+	{
+		string = strdup(pid == 0 ? "System Idle Process" : "System");
+		return string;
+	}
+	
+	/* Get process handle */
+	HANDLE hProcess = OpenProcess(PROCESS_QUERY_INFORMATION | PROCESS_VM_READ, FALSE, pid);
+	if (hProcess != NULL)
+	{
+		/* Get full Windows kernel path for the process */
+		if (GetProcessImageFileName(hProcess, read_buff, OS_MAXSTR))
+		{
+			/* Get only the process name from the string */
+			ptr = strrchr(read_buff, '\\');
+			if (ptr)
+			{
+				int len = (strlen(read_buff) - (ptr - read_buff + 1));
+				memcpy(read_buff, &(read_buff[ptr - read_buff + 1]), len);
+				read_buff[len] = '\0';
+			}
+			
+			/* Duplicate string */
+			string = strdup(read_buff);
+		} else {
+			mtwarn(WM_SYS_LOGTAG, "Unable to retrieve name for process with PID '%lu'. Error '%ld'.", pid, GetLastError());
+			string = strdup("unknown");
+		}
+		
+		/* Close process handle */
+		CloseHandle(hProcess);
+	} else {
+		mtwarn(WM_SYS_LOGTAG, "Unable to retrieve handle for process with PID '%lu'. Error '%ld'.", pid, GetLastError());
```

Comment:
This warning message is shown several times every execution with the error code 5 (access denied) when retrieving processes information. It could be interesting to find a workaround for this to avoid getting warning messages by default every scan.

Reply:
Fixed in 06ee79a0478acc8fae979ec693dee6916182f9a9.